From dd5b473e1c87fc57e58a144a52ef132ce89c1436 Mon Sep 17 00:00:00 2001
From: Prakash Brahmajyosyula <bprakash@marvell.com>
Date: Wed, 13 Feb 2019 16:50:53 +0530
Subject: [PATCH 1006/1051] octeontx2-af: NIX debugfs clean up.

This patch modifies nix debugfs to use seq_fs and seq_* apis.
There is change in usage {cq/rq/sq}_ctx dump, wrting to related debug files
maintain state and reading corresponding filedumps the context.

Below are the command and example to dump nix {cq/rq/sq} context.
Usage: echo <nixlf> [cq number/all] > cq_ctx
commands: echo 0 all > cq_ctx
	  cat cq_ctx /* to dump the cq context */
likewise we can dump context for  rq and cq also.

Change-Id: I7a5cb122aea54c002506e189fcc58df6e8f2dd25
Signed-off-by: Prakash Brahmajyosyula <bprakash@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/3954
Tested-by: sa_ip-sw-jenkins
Reviewed-by: Sunil Kovvuri Goutham <Sunil.Goutham@cavium.com>
[Kevin: The original patch got from Marvell sdk10.0_19.06]
Signed-off-by: Kevin Hao <kexin.hao@windriver.com>
---
 .../net/ethernet/marvell/octeontx2/af/rvu.h   |   3 +
 .../marvell/octeontx2/af/rvu_debugfs.c        | 539 ++++++++++--------
 2 files changed, 317 insertions(+), 225 deletions(-)

diff --git a/drivers/net/ethernet/marvell/octeontx2/af/rvu.h b/drivers/net/ethernet/marvell/octeontx2/af/rvu.h
index 14e3470a20f5..269fb6588600 100644
--- a/drivers/net/ethernet/marvell/octeontx2/af/rvu.h
+++ b/drivers/net/ethernet/marvell/octeontx2/af/rvu.h
@@ -56,6 +56,9 @@ struct rvu_debugfs {
 	struct dentry *sso_hws;
 	struct dump_ctx npa_aura_ctx;
 	struct dump_ctx npa_pool_ctx;
+	struct dump_ctx nix_cq_ctx;
+	struct dump_ctx nix_rq_ctx;
+	struct dump_ctx nix_sq_ctx;
 	int npa_qsize_id;
 };
 #endif
diff --git a/drivers/net/ethernet/marvell/octeontx2/af/rvu_debugfs.c b/drivers/net/ethernet/marvell/octeontx2/af/rvu_debugfs.c
index 01b4c70f85fd..37cb70b58db3 100644
--- a/drivers/net/ethernet/marvell/octeontx2/af/rvu_debugfs.c
+++ b/drivers/net/ethernet/marvell/octeontx2/af/rvu_debugfs.c
@@ -841,215 +841,228 @@ static void rvu_dbg_cgx_init(struct rvu *rvu)
 }
 
 /* Dumps given nix_sq's context */
-static void print_nix_sq_ctx(struct nix_aq_enq_rsp *rsp)
-{
-	struct  nix_sq_ctx_s *sq_ctx = &rsp->sq;
-
-	pr_info("W0: sqe_way_mask \t\t%d\nW0: cq \t\t\t\t%d\n",
-		sq_ctx->sqe_way_mask, sq_ctx->cq);
-	pr_info("W0: sdp_mcast \t\t\t%d\nW0: substream \t\t\t0x%03x\n",
-		sq_ctx->sdp_mcast, sq_ctx->substream);
-	pr_info("W0: qint_idx \t\t\t%d\nW0: ena \t\t\t\t%d\n\n",
-		sq_ctx->qint_idx, sq_ctx->ena);
-
-	pr_info("W1: sqb_count \t\t\t%d\nW1: default_chan \t\t%d\n",
-		sq_ctx->sqb_count, sq_ctx->default_chan);
-	pr_info("W1: smq_rr_quantum \t\t%d\nW1: sso_ena \t\t\t%d\n",
-		sq_ctx->smq_rr_quantum, sq_ctx->sso_ena);
-	pr_info("W1: xoff \t\t\t%d\nW1: cq_ena \t\t\t%d\nW1: smq\t\t\t\t%d\n\n",
-		sq_ctx->xoff, sq_ctx->cq_ena, sq_ctx->smq);
-
-	pr_info("W2: sqe_stype \t\t\t%d\nW2: sq_int_ena \t\t\t%d\n",
-		sq_ctx->sqe_stype, sq_ctx->sq_int_ena);
-	pr_info("W2: sq_int  \t\t\t%d\nW2: sqb_aura \t\t\t%d\n",
-		sq_ctx->sq_int, sq_ctx->sqb_aura);
-	pr_info("W2: smq_rr_count \t\t%d\n\n",  sq_ctx->smq_rr_count);
-
-	pr_info("W3: smq_next_sq_vld\t\t%d\nW3: smq_pend\t\t\t%d\n",
-		sq_ctx->smq_next_sq_vld, sq_ctx->smq_pend);
-	pr_info("W3: smenq_next_sqb_vld  \t\t%d\nW3: head_offset\t\t\t%d\n",
-		sq_ctx->smenq_next_sqb_vld, sq_ctx->head_offset);
-	pr_info("W3: smenq_offset\t\t\t%d\nW3: tail_offset \t\t\t%d\n",
-		sq_ctx->smenq_offset, sq_ctx->tail_offset);
-	pr_info("W3: smq_lso_segnum \t\t%d\nW3: smq_next_sq \t\t\t%d\n",
-		sq_ctx->smq_lso_segnum, sq_ctx->smq_next_sq);
-	pr_info("W3: mnq_dis \t\t\t%d\nW3: lmt_dis \t\t\t%d\n",
-		sq_ctx->mnq_dis, sq_ctx->lmt_dis);
-	pr_info("W3: cq_limit\t\t\t%d\nW3: max_sqe_size\t\t\t%d\n\n",
-		sq_ctx->cq_limit, sq_ctx->max_sqe_size);
-
-	pr_info("W4: next_sqb \t\t\t%llx\n\n", sq_ctx->next_sqb);
-	pr_info("W5: tail_sqb \t\t\t%llx\n\n", sq_ctx->tail_sqb);
-	pr_info("W6: smenq_sqb \t\t\t%llx\n\n", sq_ctx->smenq_sqb);
-	pr_info("W7: smenq_next_sqb \t\t%llx\n\n", sq_ctx->smenq_next_sqb);
-	pr_info("W8: head_sqb \t\t\t%llx\n\n", sq_ctx->head_sqb);
-
-	pr_info("W9: vfi_lso_vld \t\t\t%d\nW9: vfi_lso_vlan1_ins_ena\t%d\n",
-		sq_ctx->vfi_lso_vld, sq_ctx->vfi_lso_vlan1_ins_ena);
-	pr_info("W9: vfi_lso_vlan0_ins_ena\t%d\nW9: vfi_lso_mps\t\t\t%d\n",
-		sq_ctx->vfi_lso_vlan0_ins_ena, sq_ctx->vfi_lso_mps);
-	pr_info("W9: vfi_lso_sb\t\t\t%d\nW9: vfi_lso_sizem1\t\t%d\n",
-		sq_ctx->vfi_lso_sb, sq_ctx->vfi_lso_sizem1);
-	pr_info("W9: vfi_lso_total\t\t%d\n\n",  sq_ctx->vfi_lso_total);
-
-	pr_info("W10: scm_lso_rem  \t\t%llu\n\n", (u64)sq_ctx->scm_lso_rem);
-	pr_info("W11: octs \t\t\t%llu\n\n", (u64)sq_ctx->octs);
-	pr_info("W12: pkts \t\t\t%llu\n\n", (u64)sq_ctx->pkts);
-	pr_info("W14: dropped_octs \t\t%llu\n\n", (u64)sq_ctx->dropped_octs);
-	pr_info("W15: dropped_pkts \t\t%llu\n\n", (u64)sq_ctx->dropped_pkts);
+static void print_nix_sq_ctx(struct seq_file *m, struct nix_aq_enq_rsp *rsp)
+{
+	struct nix_sq_ctx_s *sq_ctx = &rsp->sq;
+
+	seq_printf(m, "W0: sqe_way_mask \t\t%d\nW0: cq \t\t\t\t%d\n",
+		   sq_ctx->sqe_way_mask, sq_ctx->cq);
+	seq_printf(m, "W0: sdp_mcast \t\t\t%d\nW0: substream \t\t\t0x%03x\n",
+		   sq_ctx->sdp_mcast, sq_ctx->substream);
+	seq_printf(m, "W0: qint_idx \t\t%d\nW0: ena \t\t\t\t%d\n\n",
+		   sq_ctx->qint_idx, sq_ctx->ena);
+
+	seq_printf(m, "W1: sqb_count \t\t\t%d\nW1: default_chan \t\t%d\n",
+		   sq_ctx->sqb_count, sq_ctx->default_chan);
+	seq_printf(m, "W1: smq_rr_quantum \t\t%d\nW1: sso_ena \t\t\t%d\n",
+		   sq_ctx->smq_rr_quantum, sq_ctx->sso_ena);
+	seq_printf(m, "W1: xoff \t\t\t%d\nW1: cq_ena \t\t\t%d\nW1: smq\t\t\t\t%d\n\n",
+		   sq_ctx->xoff, sq_ctx->cq_ena, sq_ctx->smq);
+
+	seq_printf(m, "W2: sqe_stype \t\t\t%d\nW2: sq_int_ena \t\t\t%d\n",
+		   sq_ctx->sqe_stype, sq_ctx->sq_int_ena);
+	seq_printf(m, "W2: sq_int \t\t\t%d\nW2: sqb_aura \t\t\t%d\n",
+		   sq_ctx->sq_int, sq_ctx->sqb_aura);
+	seq_printf(m, "W2: smq_rr_count \t\t%d\n\n", sq_ctx->smq_rr_count);
+
+	seq_printf(m, "W3: smq_next_sq_vld\t\t%d\nW3: smq_pend\t\t\t%d\n",
+		   sq_ctx->smq_next_sq_vld, sq_ctx->smq_pend);
+	seq_printf(m, "W3: smenq_next_sqb_vld \t\t%d\nW3: head_offset\t\t\t%d\n",
+		   sq_ctx->smenq_next_sqb_vld, sq_ctx->head_offset);
+	seq_printf(m, "W3: smenq_offset\t\t%d\nW3: tail_offset\t\t%d\n",
+		   sq_ctx->smenq_offset, sq_ctx->tail_offset);
+	seq_printf(m, "W3: smq_lso_segnum \t\t%d\nW3: smq_next_sq\t\t\t%d\n",
+		   sq_ctx->smq_lso_segnum, sq_ctx->smq_next_sq);
+	seq_printf(m, "W3: mnq_dis \t\t\t%d\nW3: lmt_dis \t\t\t%d\n",
+		   sq_ctx->mnq_dis, sq_ctx->lmt_dis);
+	seq_printf(m, "W3: cq_limit\t\t\t%d\nW3: max_sqe_size\t\t%d\n\n",
+		   sq_ctx->cq_limit, sq_ctx->max_sqe_size);
+
+	seq_printf(m, "W4: next_sqb \t\t\t%llx\n\n", sq_ctx->next_sqb);
+	seq_printf(m, "W5: tail_sqb \t\t\t%llx\n\n", sq_ctx->tail_sqb);
+	seq_printf(m, "W6: smenq_sqb \t\t\t%llx\n\n", sq_ctx->smenq_sqb);
+	seq_printf(m, "W7: smenq_next_sqb \t\t%llx\n\n",
+		   sq_ctx->smenq_next_sqb);
+
+	seq_printf(m, "W8: head_sqb\t\t\t%llx\n\n", sq_ctx->head_sqb);
+
+	seq_printf(m, "W9: vfi_lso_vld\t\t\t%d\nW9: vfi_lso_vlan1_ins_ena\t%d\n",
+		   sq_ctx->vfi_lso_vld, sq_ctx->vfi_lso_vlan1_ins_ena);
+	seq_printf(m, "W9: vfi_lso_vlan0_ins_ena\t%d\nW9: vfi_lso_mps\t\t\t%d\n",
+		   sq_ctx->vfi_lso_vlan0_ins_ena, sq_ctx->vfi_lso_mps);
+	seq_printf(m, "W9: vfi_lso_sb\t\t\t%d\nW9: vfi_lso_sizem1\t\t%d\n",
+		   sq_ctx->vfi_lso_sb, sq_ctx->vfi_lso_sizem1);
+	seq_printf(m, "W9: vfi_lso_total\t\t%d\n\n", sq_ctx->vfi_lso_total);
+
+	seq_printf(m, "W10: scm_lso_rem \t\t%llu\n\n",
+		   (u64)sq_ctx->scm_lso_rem);
+	seq_printf(m, "W11: octs \t\t\t%llu\n\n", (u64)sq_ctx->octs);
+	seq_printf(m, "W12: pkts \t\t\t%llu\n\n", (u64)sq_ctx->pkts);
+	seq_printf(m, "W14: dropped_octs \t\t%llu\n\n",
+		   (u64)sq_ctx->dropped_octs);
+	seq_printf(m, "W15: dropped_pkts \t\t%llu\n\n",
+		   (u64)sq_ctx->dropped_pkts);
 }
 
 /* Dumps given nix_rq's context */
-static void print_nix_rq_ctx(struct nix_aq_enq_rsp *rsp)
+static void print_nix_rq_ctx(struct seq_file *m, struct nix_aq_enq_rsp *rsp)
 {
 	struct  nix_rq_ctx_s *rq_ctx = &rsp->rq;
 
-	pr_info("W0: wqe_aura \t\t\t%d\nW0: substream \t\t\t0x%03x\n",
-		rq_ctx->wqe_aura, rq_ctx->substream);
-	pr_info("W0: cq \t\t\t\t%d\nW0: ena_wqwd \t\t\t%d\n",
-		rq_ctx->cq, rq_ctx->ena_wqwd);
-	pr_info("W0: ipsech_ena \t\t\t%d\nW0: sso_ena \t\t\t%d\n",
-		rq_ctx->ipsech_ena, rq_ctx->sso_ena);
-	pr_info("W0: ena \t\t\t\t%d\n\n", rq_ctx->ena);
-
-	pr_info("W1: lpb_drop_ena \t\t%d\nW1: spb_drop_ena \t\t%d\n",
-		rq_ctx->lpb_drop_ena, rq_ctx->spb_drop_ena);
-	pr_info("W1: xqe_drop_ena \t\t%d\nW1: wqe_caching \t\t\t%d\n",
-		rq_ctx->xqe_drop_ena, rq_ctx->wqe_caching);
-	pr_info("W1: pb_caching \t\t\t%d\nW1: sso_tt \t\t\t%d\n",
-		rq_ctx->pb_caching, rq_ctx->sso_tt);
-	pr_info("W1: sso_grp \t\t\t%d\nW1: lpb_aura \t\t\t%d\n",
-		rq_ctx->sso_grp, rq_ctx->lpb_aura);
-	pr_info("W1: spb_aura \t\t\t%d\n\n", rq_ctx->spb_aura);
-
-	pr_info("W2: xqe_hdr_split \t\t%d\nW2: xqe_imm_copy \t\t%d\n",
-		rq_ctx->xqe_hdr_split, rq_ctx->xqe_imm_copy);
-	pr_info("W2: xqe_imm_size \t\t%d\nW2: later_skip \t\t\t%d\n",
-		rq_ctx->xqe_imm_size, rq_ctx->later_skip);
-	pr_info("W2: first_skip \t\t\t%d\nW2: lpb_sizem1 \t\t\t%d\n",
-		rq_ctx->first_skip, rq_ctx->lpb_sizem1);
-	pr_info("W2: spb_ena \t\t\t%d\nW2: wqe_skip \t\t\t%d\n",
-		rq_ctx->spb_ena, rq_ctx->wqe_skip);
-	pr_info("W2: spb_sizem1 \t\t\t%d\n\n", rq_ctx->spb_sizem1);
-
-	pr_info("W3: spb_pool_pass \t\t%d\nW3: spb_pool_drop \t\t%d\n",
-		rq_ctx->spb_pool_pass, rq_ctx->spb_pool_drop);
-	pr_info("W3: spb_aura_pass \t\t%d\nW3: spb_aura_drop \t\t%d\n",
-		rq_ctx->spb_aura_pass, rq_ctx->spb_aura_drop);
-	pr_info("W3: wqe_pool_pass \t\t%d\nW3: wqe_pool_drop \t\t%d\n",
-		rq_ctx->wqe_pool_pass, rq_ctx->wqe_pool_drop);
-	pr_info("W3: xqe_pass \t\t\t%d\nW3: xqe_drop \t\t\t%d\n\n",
-		rq_ctx->xqe_pass, rq_ctx->xqe_drop);
-
-	pr_info("W4: qint_idx \t\t\t%d\nW4: rq_int_ena \t\t\t%d\n",
-		rq_ctx->qint_idx, rq_ctx->rq_int_ena);
-	pr_info("W4: rq_int \t\t\t%d\nW4: lpb_pool_pass \t\t%d\n",
-		rq_ctx->rq_int, rq_ctx->lpb_pool_pass);
-	pr_info("W4: lpb_pool_drop \t\t%d\nW4: lpb_aura_pass \t\t%d\n",
-		rq_ctx->lpb_pool_drop, rq_ctx->lpb_aura_pass);
-	pr_info("W4: lpb_aura_drop \t\t%d\n\n", rq_ctx->lpb_aura_drop);
-
-	pr_info("W5: flow_tagw \t\t\t%d\nW5: bad_utag \t\t\t%d\n",
-		rq_ctx->flow_tagw, rq_ctx->bad_utag);
-	pr_info("W5: good_utag \t\t\t%d\nW5: ltag \t\t\t%d\n\n",
-		rq_ctx->good_utag, rq_ctx->ltag);
-
-	pr_info("W6: octs \t\t\t%llu\n\n", (u64)rq_ctx->octs);
-	pr_info("W7: pkts \t\t\t%llu\n\n", (u64)rq_ctx->pkts);
-	pr_info("W8: drop_octs \t\t\t%llu\n\n", (u64)rq_ctx->drop_octs);
-	pr_info("W9: drop_pkts \t\t\t%llu\n\n", (u64)rq_ctx->drop_pkts);
-	pr_info("W10: re_pkts \t\t\t%llu\n", (u64)rq_ctx->re_pkts);
+	seq_printf(m, "W0: wqe_aura \t\t\t%d\nW0: substream \t\t\t0x%03x\n",
+		   rq_ctx->wqe_aura, rq_ctx->substream);
+	seq_printf(m, "W0: cq \t\t\t\t%d\nW0: ena_wqwd \t\t\t%d\n",
+		   rq_ctx->cq, rq_ctx->ena_wqwd);
+	seq_printf(m, "W0: ipsech_ena \t\t\t%d\nW0: sso_ena \t\t\t%d\n",
+		   rq_ctx->ipsech_ena, rq_ctx->sso_ena);
+	seq_printf(m, "W0: ena \t\t\t\t%d\n\n", rq_ctx->ena);
+
+	seq_printf(m, "W1: lpb_drop_ena \t\t%d\nW1: spb_drop_ena \t\t%d\n",
+		   rq_ctx->lpb_drop_ena, rq_ctx->spb_drop_ena);
+	seq_printf(m, "W1: xqe_drop_ena \t\t%d\nW1: wqe_caching \t\t\t%d\n",
+		   rq_ctx->xqe_drop_ena, rq_ctx->wqe_caching);
+	seq_printf(m, "W1: pb_caching \t\t\t%d\nW1: sso_tt \t\t\t%d\n",
+		   rq_ctx->pb_caching, rq_ctx->sso_tt);
+	seq_printf(m, "W1: sso_grp \t\t\t%d\nW1: lpb_aura \t\t\t%d\n",
+		   rq_ctx->sso_grp, rq_ctx->lpb_aura);
+	seq_printf(m, "W1: spb_aura \t\t\t%d\n\n", rq_ctx->spb_aura);
+
+	seq_printf(m, "W2: xqe_hdr_split \t\t%d\nW2: xqe_imm_copy \t\t%d\n",
+		   rq_ctx->xqe_hdr_split, rq_ctx->xqe_imm_copy);
+	seq_printf(m, "W2: xqe_imm_size \t\t%d\nW2: later_skip \t\t\t%d\n",
+		   rq_ctx->xqe_imm_size, rq_ctx->later_skip);
+	seq_printf(m, "W2: first_skip \t\t\t%d\nW2: lpb_sizem1 \t\t\t%d\n",
+		   rq_ctx->first_skip, rq_ctx->lpb_sizem1);
+	seq_printf(m, "W2: spb_ena \t\t\t%d\nW2: wqe_skip \t\t\t%d\n",
+		   rq_ctx->spb_ena, rq_ctx->wqe_skip);
+	seq_printf(m, "W2: spb_sizem1 \t\t\t%d\n\n", rq_ctx->spb_sizem1);
+
+	seq_printf(m, "W3: spb_pool_pass \t\t%d\nW3: spb_pool_drop \t\t%d\n",
+		   rq_ctx->spb_pool_pass, rq_ctx->spb_pool_drop);
+	seq_printf(m, "W3: spb_aura_pass \t\t%d\nW3: spb_aura_drop \t\t%d\n",
+		   rq_ctx->spb_aura_pass, rq_ctx->spb_aura_drop);
+	seq_printf(m, "W3: wqe_pool_pass \t\t%d\nW3: wqe_pool_drop \t\t%d\n",
+		   rq_ctx->wqe_pool_pass, rq_ctx->wqe_pool_drop);
+	seq_printf(m, "W3: xqe_pass \t\t\t%d\nW3: xqe_drop \t\t\t%d\n\n",
+		   rq_ctx->xqe_pass, rq_ctx->xqe_drop);
+
+	seq_printf(m, "W4: qint_idx \t\t\t%d\nW4: rq_int_ena \t\t\t%d\n",
+		   rq_ctx->qint_idx, rq_ctx->rq_int_ena);
+	seq_printf(m, "W4: rq_int \t\t\t%d\nW4: lpb_pool_pass \t\t%d\n",
+		   rq_ctx->rq_int, rq_ctx->lpb_pool_pass);
+	seq_printf(m, "W4: lpb_pool_drop \t\t%d\nW4: lpb_aura_pass \t\t%d\n",
+		   rq_ctx->lpb_pool_drop, rq_ctx->lpb_aura_pass);
+	seq_printf(m, "W4: lpb_aura_drop \t\t%d\n\n", rq_ctx->lpb_aura_drop);
+
+	seq_printf(m, "W5: flow_tagw \t\t\t%d\nW5: bad_utag \t\t\t%d\n",
+		   rq_ctx->flow_tagw, rq_ctx->bad_utag);
+	seq_printf(m, "W5: good_utag \t\t\t%d\nW5: ltag \t\t\t%d\n\n",
+		   rq_ctx->good_utag, rq_ctx->ltag);
+
+	seq_printf(m, "W6: octs \t\t\t%llu\n\n", (u64)rq_ctx->octs);
+	seq_printf(m, "W7: pkts \t\t\t%llu\n\n", (u64)rq_ctx->pkts);
+	seq_printf(m, "W8: drop_octs \t\t\t%llu\n\n", (u64)rq_ctx->drop_octs);
+	seq_printf(m, "W9: drop_pkts \t\t\t%llu\n\n", (u64)rq_ctx->drop_pkts);
+	seq_printf(m, "W10: re_pkts \t\t\t%llu\n", (u64)rq_ctx->re_pkts);
 }
 
 /* Dumps given nix_cq's context */
-static void print_nix_cq_ctx(struct nix_aq_enq_rsp *rsp)
+static void print_nix_cq_ctx(struct seq_file *m, struct nix_aq_enq_rsp *rsp)
 {
 	struct  nix_cq_ctx_s *cq_ctx = &rsp->cq;
 
-	pr_info("W0: base \t\t\t%llx\n\n", cq_ctx->base);
-
-	pr_info("W1: wrptr \t\t\t%llx\n", (u64)cq_ctx->wrptr);
-	pr_info("W1: avg_con \t\t\t%d\nW1: cint_idx \t\t\t%d\n",
-		cq_ctx->avg_con, cq_ctx->cint_idx);
-	pr_info("W1: cq_err \t\t\t%d\nW1: qint_idx \t\t\t%d\n",
-		cq_ctx->cq_err, cq_ctx->qint_idx);
-	pr_info("W1: bpid  \t\t\t%d\nW1: bp_ena \t\t\t%d\n\n",
-		cq_ctx->bpid, cq_ctx->bp_ena);
-
-	pr_info("W2: update_time \t\t\t%d\nW2:avg_level \t\t\t%d\n",
-		cq_ctx->update_time, cq_ctx->avg_level);
-	pr_info("W2: head \t\t\t%d\nW2:tail \t\t\t\t%d\n\n",
-		cq_ctx->head, cq_ctx->tail);
-
-	pr_info("W3: cq_err_int_ena \t\t%d\nW3:cq_err_int \t\t\t%d\n",
-		cq_ctx->cq_err_int_ena, cq_ctx->cq_err_int);
-	pr_info("W3: qsize \t\t\t%d\nW3:caching \t\t\t%d\n",
-		cq_ctx->qsize, cq_ctx->caching);
-	pr_info("W3: substream \t\t\t0x%03x\nW3: ena \t\t\t\t%d\n",
-		cq_ctx->substream, cq_ctx->ena);
-	pr_info("W3: drop_ena \t\t\t%d\nW3: drop \t\t\t%d\n",
-		cq_ctx->drop_ena, cq_ctx->drop);
-	pr_info("W3: dp \t\t\t\t%d\n\n", cq_ctx->dp);
-}
-
-static void read_nix_ctx(struct rvu *rvu, bool all, int nixlf,
-			 int id, int ctype, char *ctype_string)
-{
-	void (*print_nix_ctx)(struct nix_aq_enq_rsp *rsp) = NULL;
-	int blkaddr, qidx, rc, max_id = 0;
-	struct rvu_hwinfo *hw = rvu->hw;
+	seq_printf(m, "W0: base \t\t\t%llx\n\n", cq_ctx->base);
+
+	seq_printf(m, "W1: wrptr \t\t\t%llx\n", (u64)cq_ctx->wrptr);
+	seq_printf(m, "W1: avg_con \t\t\t%d\nW1: cint_idx \t\t\t%d\n",
+		   cq_ctx->avg_con, cq_ctx->cint_idx);
+	seq_printf(m, "W1: cq_err \t\t\t%d\nW1: qint_idx \t\t\t%d\n",
+		   cq_ctx->cq_err, cq_ctx->qint_idx);
+	seq_printf(m, "W1: bpid \t\t\t%d\nW1: bp_ena \t\t\t%d\n\n",
+		   cq_ctx->bpid, cq_ctx->bp_ena);
+
+	seq_printf(m, "W2: update_time \t\t\t%d\nW2:avg_level \t\t\t%d\n",
+		   cq_ctx->update_time, cq_ctx->avg_level);
+	seq_printf(m, "W2: head \t\t\t%d\nW2:tail \t\t\t\t%d\n\n",
+		   cq_ctx->head, cq_ctx->tail);
+
+	seq_printf(m, "W3: cq_err_int_ena \t\t%d\nW3:cq_err_int \t\t\t%d\n",
+		   cq_ctx->cq_err_int_ena, cq_ctx->cq_err_int);
+	seq_printf(m, "W3: qsize \t\t\t%d\nW3:caching \t\t\t%d\n",
+		   cq_ctx->qsize, cq_ctx->caching);
+	seq_printf(m, "W3: substream \t\t\t0x%03x\nW3: ena \t\t\t\t%d\n",
+		   cq_ctx->substream, cq_ctx->ena);
+	seq_printf(m, "W3: drop_ena \t\t\t%d\nW3: drop \t\t\t%d\n",
+		   cq_ctx->drop_ena, cq_ctx->drop);
+	seq_printf(m, "W3: dp \t\t\t\t%d\n\n", cq_ctx->dp);
+}
+
+static int rvu_dbg_nix_queue_ctx_display(struct seq_file *filp,
+					 void *unused, int ctype)
+{
+	void (*print_nix_ctx)(struct seq_file *filp,
+			      struct nix_aq_enq_rsp *rsp) = NULL;
+	struct rvu *rvu = filp->private;
 	struct nix_aq_enq_req aq_req;
 	struct nix_aq_enq_rsp rsp;
-	struct rvu_block *block;
+	char *ctype_string = NULL;
+	int qidx, rc, max_id = 0;
 	struct rvu_pfvf *pfvf;
-	u64 pcifunc;
+	int nixlf, id, all;
+	u16 pcifunc;
 
-	blkaddr = rvu_get_blkaddr(rvu, BLKTYPE_NIX, 0);
-	if (blkaddr < 0)
-		return;
+	switch (ctype) {
+	case NIX_AQ_CTYPE_CQ:
+		nixlf = rvu->rvu_dbg.nix_cq_ctx.lf;
+		id = rvu->rvu_dbg.nix_cq_ctx.id;
+		all = rvu->rvu_dbg.nix_cq_ctx.all;
+		break;
 
-	block = &hw->block[blkaddr];
-	if (nixlf < 0 || nixlf >= block->lf.max) {
-		pr_info("Invalid NIX LF, Valid range 0-%d\n",
-			block->lf.max - 1);
-		return;
-	}
+	case NIX_AQ_CTYPE_SQ:
+		nixlf = rvu->rvu_dbg.nix_sq_ctx.lf;
+		id = rvu->rvu_dbg.nix_sq_ctx.id;
+		all = rvu->rvu_dbg.nix_sq_ctx.all;
+		break;
 
-	pcifunc = block->fn_map[nixlf];
-	if (!pcifunc) {
-		pr_info("This NIX LF is not attached to any RVU PFFUNC\n");
-		return;
+	case NIX_AQ_CTYPE_RQ:
+		nixlf = rvu->rvu_dbg.nix_rq_ctx.lf;
+		id = rvu->rvu_dbg.nix_rq_ctx.id;
+		all = rvu->rvu_dbg.nix_rq_ctx.all;
+		break;
+
+	default:
+		return -EINVAL;
 	}
 
+	if (!rvu_dbg_is_valid_lf(rvu, BLKTYPE_NIX, nixlf, &pcifunc))
+		return -EINVAL;
+
 	pfvf = rvu_get_pfvf(rvu, pcifunc);
 	if (ctype == NIX_AQ_CTYPE_SQ && !pfvf->sq_ctx) {
-		pr_info("SQ context is not initialized\n");
-		return;
+		seq_puts(filp, "SQ context is not initialized\n");
+		return -EINVAL;
 	} else if (ctype == NIX_AQ_CTYPE_RQ && !pfvf->rq_ctx) {
-		pr_info("RQ context is not initialized\n");
-		return;
+		seq_puts(filp, "RQ context is not initialized\n");
+		return -EINVAL;
 	} else if (ctype == NIX_AQ_CTYPE_CQ && !pfvf->cq_ctx) {
-		pr_info("CQ context is not initialized\n");
-		return;
+		seq_puts(filp, "CQ context is not initialized\n");
+		return -EINVAL;
 	}
 
 	if (ctype == NIX_AQ_CTYPE_SQ) {
-		max_id =  pfvf->sq_ctx->qsize;
+		max_id = pfvf->sq_ctx->qsize;
+		ctype_string = "sq";
 		print_nix_ctx = print_nix_sq_ctx;
 	} else if (ctype == NIX_AQ_CTYPE_RQ) {
-		max_id =  pfvf->rq_ctx->qsize;
+		max_id = pfvf->rq_ctx->qsize;
+		ctype_string = "rq";
 		print_nix_ctx = print_nix_rq_ctx;
 	} else if (ctype == NIX_AQ_CTYPE_CQ) {
-		max_id =  pfvf->cq_ctx->qsize;
+		max_id = pfvf->cq_ctx->qsize;
+		ctype_string = "cq";
 		print_nix_ctx = print_nix_cq_ctx;
 	}
 
-	if (id < 0 || id >= max_id) {
-		pr_info("Invalid %s_ctx valid range 0-%d\n",
-			ctype_string, max_id - 1);
-		return;
-	}
-
 	memset(&aq_req, 0, sizeof(struct nix_aq_enq_req));
 	aq_req.hdr.pcifunc = pcifunc;
 	aq_req.ctype = ctype;
@@ -1060,28 +1073,94 @@ static void read_nix_ctx(struct rvu *rvu, bool all, int nixlf,
 		max_id = id + 1;
 	for (qidx = id; qidx < max_id; qidx++) {
 		aq_req.qidx = qidx;
-		pr_info("=====%s_ctx for nixlf:%d and qidx:%d is=====\n",
-			ctype_string, nixlf, aq_req.qidx);
+		seq_printf(filp, "=====%s_ctx for nixlf:%d and qidx:%d is=====\n",
+			   ctype_string, nixlf, aq_req.qidx);
 		rc = rvu_mbox_handler_nix_aq_enq(rvu, &aq_req, &rsp);
 		if (rc) {
-			pr_info("Failed to read the context\n");
+			seq_puts(filp, "Failed to read the context\n");
+			return -EINVAL;
+		}
+		print_nix_ctx(filp, &rsp);
+	}
+	return 0;
+}
+
+static void write_nix_queue_ctx(struct rvu *rvu, bool all, int nixlf,
+				int id, int ctype, char *ctype_string)
+{
+	struct rvu_pfvf *pfvf;
+	int max_id = 0;
+	u16 pcifunc;
+
+	if (!rvu_dbg_is_valid_lf(rvu, BLKTYPE_NIX, nixlf, &pcifunc))
+		return;
+
+	pfvf = rvu_get_pfvf(rvu, pcifunc);
+
+	if (ctype == NIX_AQ_CTYPE_SQ) {
+		if (!pfvf->sq_ctx) {
+			dev_warn(rvu->dev, "SQ context is not initialized\n");
+			return;
+		}
+		max_id = pfvf->sq_ctx->qsize;
+	} else if (ctype == NIX_AQ_CTYPE_RQ) {
+		if (!pfvf->rq_ctx) {
+			dev_warn(rvu->dev, "RQ context is not initialized\n");
 			return;
 		}
-		print_nix_ctx(&rsp);
+		max_id = pfvf->rq_ctx->qsize;
+	} else if (ctype == NIX_AQ_CTYPE_CQ) {
+		if (!pfvf->cq_ctx) {
+			dev_warn(rvu->dev, "CQ context is not initialized\n");
+			return;
+		}
+		max_id = pfvf->cq_ctx->qsize;
+	}
+
+	if (id < 0 || id >= max_id) {
+		dev_warn(rvu->dev, "Invalid %s_ctx valid range 0-%d\n",
+			 ctype_string, max_id - 1);
+		return;
+	}
+	switch (ctype) {
+	case NIX_AQ_CTYPE_CQ:
+		rvu->rvu_dbg.nix_cq_ctx.lf = nixlf;
+		rvu->rvu_dbg.nix_cq_ctx.id = id;
+		rvu->rvu_dbg.nix_cq_ctx.all = all;
+		break;
+
+	case NIX_AQ_CTYPE_SQ:
+		rvu->rvu_dbg.nix_sq_ctx.lf = nixlf;
+		rvu->rvu_dbg.nix_sq_ctx.id = id;
+		rvu->rvu_dbg.nix_sq_ctx.all = all;
+		break;
+
+	case NIX_AQ_CTYPE_RQ:
+		rvu->rvu_dbg.nix_rq_ctx.lf = nixlf;
+		rvu->rvu_dbg.nix_rq_ctx.id = id;
+		rvu->rvu_dbg.nix_rq_ctx.all = all;
+		break;
+
+	default:
+		return;
 	}
+	return;
 }
 
-static ssize_t rvu_dbg_nix_ctx_display(struct file *filp,
-				       const char __user *buffer,
-				       size_t count, loff_t *ppos, int ctype)
+static ssize_t rvu_dbg_nix_queue_ctx_write(struct file *filp,
+					   const char __user *buffer,
+					   size_t count, loff_t *ppos,
+					   int ctype)
 {
-	struct rvu *rvu = filp->private_data;
+	struct seq_file *m = filp->private_data;
+	struct rvu *rvu = m->private;
 	char *cmd_buf, *ctype_string;
 	int nixlf, id = 0;
 	bool all = false;
 
 	if ((*ppos != 0) || !count)
 		return 0;
+
 	switch (ctype) {
 	case NIX_AQ_CTYPE_SQ:
 		ctype_string = "sq";
@@ -1099,87 +1178,97 @@ static ssize_t rvu_dbg_nix_ctx_display(struct file *filp,
 
 	if (!cmd_buf)
 		return count;
+
 	if (parse_cmd_buffer_ctx(cmd_buf, &count, buffer,
 				 &nixlf, &id, &all) < 0) {
-		pr_info("Usage: echo <nixlf> [%s number/all] > %s_ctx\n",
-			ctype_string, ctype_string);
+		dev_info(rvu->dev, "Usage: echo <nixlf> [%s number/all] > %s_ctx\n",
+			 ctype_string, ctype_string);
 	} else {
-		read_nix_ctx(rvu, all, nixlf, id, ctype, ctype_string);
+		write_nix_queue_ctx(rvu, all, nixlf, id, ctype, ctype_string);
 	}
 
 	kfree(cmd_buf);
 	return count;
 }
 
-static ssize_t rvu_dbg_nix_sq_ctx_display(struct file *filp,
-					  const char __user *buffer,
-					  size_t count, loff_t *ppos)
+static ssize_t rvu_dbg_nix_sq_ctx_write(struct file *filp,
+					const char __user *buffer,
+					size_t count, loff_t *ppos)
 {
-	return  rvu_dbg_nix_ctx_display(filp, buffer,
-					count, ppos, NIX_AQ_CTYPE_SQ);
+	return rvu_dbg_nix_queue_ctx_write(filp, buffer, count, ppos,
+					    NIX_AQ_CTYPE_SQ);
 }
 
-RVU_DEBUG_FOPS(nix_sq_ctx, NULL, nix_sq_ctx_display);
+static int rvu_dbg_nix_sq_ctx_display(struct seq_file *filp, void *unused)
+{
+	return rvu_dbg_nix_queue_ctx_display(filp, unused, NIX_AQ_CTYPE_SQ);
+}
 
-static ssize_t rvu_dbg_nix_rq_ctx_display(struct file *filp,
-					  const char __user *buffer,
-					  size_t count, loff_t *ppos)
+RVU_DEBUG_SEQ_FOPS(nix_sq_ctx, nix_sq_ctx_display, nix_sq_ctx_write);
+
+static ssize_t rvu_dbg_nix_rq_ctx_write(struct file *filp,
+					const char __user *buffer,
+					size_t count, loff_t *ppos)
 {
-	return  rvu_dbg_nix_ctx_display(filp, buffer,
-					count, ppos, NIX_AQ_CTYPE_RQ);
+	return rvu_dbg_nix_queue_ctx_write(filp, buffer, count, ppos,
+					    NIX_AQ_CTYPE_RQ);
 }
 
-RVU_DEBUG_FOPS(nix_rq_ctx, NULL, nix_rq_ctx_display);
+static int rvu_dbg_nix_rq_ctx_display(struct seq_file *filp, void  *unused)
+{
+	return rvu_dbg_nix_queue_ctx_display(filp, unused,  NIX_AQ_CTYPE_RQ);
+}
 
-static ssize_t rvu_dbg_nix_cq_ctx_display(struct file *filp,
-					  const char __user *buffer,
-					  size_t count, loff_t *ppos)
+RVU_DEBUG_SEQ_FOPS(nix_rq_ctx, nix_rq_ctx_display, nix_rq_ctx_write);
+
+static ssize_t rvu_dbg_nix_cq_ctx_write(struct file *filp,
+					const char __user *buffer,
+					size_t count, loff_t *ppos)
+{
+	return rvu_dbg_nix_queue_ctx_write(filp, buffer, count, ppos,
+					    NIX_AQ_CTYPE_CQ);
+}
+
+static int rvu_dbg_nix_cq_ctx_display(struct seq_file *filp, void *unused)
 {
-	return  rvu_dbg_nix_ctx_display(filp, buffer,
-					count, ppos, NIX_AQ_CTYPE_CQ);
+	return rvu_dbg_nix_queue_ctx_display(filp, unused, NIX_AQ_CTYPE_CQ);
 }
 
-RVU_DEBUG_FOPS(nix_cq_ctx, NULL, nix_cq_ctx_display);
+RVU_DEBUG_SEQ_FOPS(nix_cq_ctx, nix_cq_ctx_display, nix_cq_ctx_write);
 
-static ssize_t rvu_dbg_nix_ndc_rx_cache_display(struct file *filp,
-						char __user *buffer,
-						size_t count, loff_t *ppos)
+static int rvu_dbg_nix_ndc_rx_cache_display(struct seq_file *filp, void *unused)
 {
-	return ndc_blk_cache_stats(filp->private_data, NIX0_RX,
+	return ndc_blk_cache_stats(filp, NIX0_RX,
 				   BLKADDR_NDC_NIX0_RX);
 }
 
-RVU_DEBUG_FOPS(nix_ndc_rx_cache, nix_ndc_rx_cache_display, NULL);
+RVU_DEBUG_SEQ_FOPS(nix_ndc_rx_cache, nix_ndc_rx_cache_display, NULL);
 
-static ssize_t rvu_dbg_nix_ndc_tx_cache_display(struct file *filp,
-						char __user *buffer,
-						size_t count, loff_t *ppos)
+static int rvu_dbg_nix_ndc_tx_cache_display(struct seq_file *filp, void *unused)
 {
-	return ndc_blk_cache_stats(filp->private_data, NIX0_TX,
+	return ndc_blk_cache_stats(filp, NIX0_TX,
 				   BLKADDR_NDC_NIX0_TX);
 }
 
-RVU_DEBUG_FOPS(nix_ndc_tx_cache, nix_ndc_tx_cache_display, NULL);
+RVU_DEBUG_SEQ_FOPS(nix_ndc_tx_cache, nix_ndc_tx_cache_display, NULL);
 
-static ssize_t rvu_dbg_nix_ndc_rx_hits_miss_display(struct file *filp,
-						    char __user *buffer,
-						    size_t count, loff_t *ppos)
+static int rvu_dbg_nix_ndc_rx_hits_miss_display(struct seq_file *filp,
+						void *unused)
 {
-	return ndc_blk_hits_miss_stats(filp->private_data,
+	return ndc_blk_hits_miss_stats(filp,
 				      NPA0_U, BLKADDR_NDC_NIX0_RX);
 }
 
-RVU_DEBUG_FOPS(nix_ndc_rx_hits_miss, nix_ndc_rx_hits_miss_display, NULL);
+RVU_DEBUG_SEQ_FOPS(nix_ndc_rx_hits_miss, nix_ndc_rx_hits_miss_display, NULL);
 
-static ssize_t rvu_dbg_nix_ndc_tx_hits_miss_display(struct file *filp,
-						    char __user *buffer,
-						    size_t count, loff_t *ppos)
+static int rvu_dbg_nix_ndc_tx_hits_miss_display(struct seq_file *filp,
+						void *unused)
 {
-	return ndc_blk_hits_miss_stats(filp->private_data,
+	return ndc_blk_hits_miss_stats(filp,
 				      NPA0_U, BLKADDR_NDC_NIX0_TX);
 }
 
-RVU_DEBUG_FOPS(nix_ndc_tx_hits_miss, nix_ndc_tx_hits_miss_display, NULL);
+RVU_DEBUG_SEQ_FOPS(nix_ndc_tx_hits_miss, nix_ndc_tx_hits_miss_display, NULL);
 
 static ssize_t rvu_dbg_nix_tx_stall_hwissue_display(struct file *filp,
 						    char __user *buffer,
@@ -1197,7 +1286,7 @@ static void rvu_dbg_nix_init(struct rvu *rvu)
 
 	rvu->rvu_dbg.nix = debugfs_create_dir("nix", rvu->rvu_dbg.root);
 	if (!rvu->rvu_dbg.nix) {
-		pr_info("create debugfs dir failed for nix\n");
+		dev_err(rvu->dev, "create debugfs dir failed for nix\n");
 		return;
 	}
 
-- 
2.17.1

